base/reduce.jl

```diff
@@ -420,34 +416,176 @@ The default is `reduce_first(op, f(x))`.
+        return op(v1, v2)
+    end
+end
+function halves(inds)
+    n = length(inds) >> 1
+    return (inds[begin:begin+n-1], inds[begin+n:end])
+end
+# Recursively find the last non-singleton range in the tuple to halve, keeping the rest whole
+_halves(tup) = length(tup[end]) > 1 ?
+                    (map(_wholes, front(tup))..., halves(tup[end])) :
+                    (_halves(front(tup))..., _wholes(tup[end]))
+_halves(::Tuple{}) = ()
+# This unconditionally re-indexes by begin:end to ensure type stability with the halving re-index
+_wholes(r) = r[begin:end], r[begin:end]
+
+function mapreduce_pairwise(f::F, op::G, itr, init, S::Union{HasLength,HasShape}, n, state...) where {F,G}
+    if n < max(10, pairwise_blocksize(f, op))
+        return mapreduce_kernel(f, op, itr, init, S, n, state...)
+    else
+        ndiv2 = n >> 1
+        v1, s = mapreduce_pairwise(f, op, itr, init, S, ndiv2, state...)
+        v2, s = mapreduce_pairwise(f, op, itr, init, S, n-ndiv2, s)
+        return (op(v1, v2), s)
+    end
+end
+function mapreduce_pairwise(f::F, op, itr, init, S::IteratorSize, n, it) where {F}
+    if n <= max(10, pairwise_blocksize(f, op))
+        v, it = mapreduce_kernel(f, op, itr, init, S, n, it)
+        return it === nothing ? (v, nothing) : (v, it)
+    else
+        ndiv2 = n >> 1
+        v1, it = mapreduce_pairwise(f, op, itr, init, S, ndiv2, it)
+        it === nothing && return (v1, nothing)
+        v2, it = mapreduce_pairwise(f, op, itr, init, S, n-ndiv2, it)
+        v = op(v1, v2)
+        return it === nothing ? (v, nothing) : (v, it)
+    end
+end
 
-_mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f, op, A)
+"""
+    mapreduce_kernel(f, op, A, init, inds) -> v
+    mapreduce_kernel(f, op, itr, init, ::Union{HasLength, HasShape}, n, [state]) -> (v, state)
+    mapreduce_kernel(f, op, itr, init, ::IteratorSize, n, valstate) -> (v, iterate(itr, state))
+
+Perform the mapped reduction with `f` and `op` akin to a `mapfoldl` but permitting SIMD-like
+optimizations and reassociations. Corresponding to `mapreduce_pairwise`, this function uses
+three (and a half) different mechanisms for accessing elements and structuring its loop(s):
+
+* Indexing with `inds` into `A`
+    * (with a dedicated optimization for CartesianIndices)
+* Iterating exactly `n` elements from `itr` with a known length and at least `n` elements remaining
+    This method additionally takes (and returns as the second element in a tuple) the current iterator state
+* Iterating *up to* `n` elements from `itr`, but possibly stopping early
+    This method additionally takes (and returns as the second element in a tuple) the *next* `(val, state)`
+    tuple from the iterator or `nothing`
+"""
+function mapreduce_kernel(f, op, A, init, inds)
+    a1 = @inbounds A[inds[begin]]
+    v = _mapreduce_start(f, op, A, init, a1)
+    length(inds) == 1 && return v
+    @simd for i in inds[begin+1:end]
+        a = @inbounds A[i]
```

Comment:
This `@inbounds` is too unsafe, IMO. It makes a wrong `length` implementation segfault. For the most basic case (e.g. iterating over a `Vector`), the compiler should be able to remove the bounds check automatically.

Reply:
This is the status quo, it's just spread around everywhere currently: [1](https://github.com/JuliaLang/julia/blob/master/base/reduce.jl#L262) and [2](https://github.com/JuliaLang/julia/blob/eebda6c040f7285d2ee1eb69be8f5a7a203b43dc/base/reduce.jl#L439) and [3](https://github.com/JuliaLang/julia/blob/eebda6c040f7285d2ee1eb69be8f5a7a203b43dc/base/reducedim.jl#L281) and [4](https://github.com/JuliaLang/julia/blob/eebda6c040f7285d2ee1eb69be8f5a7a203b43dc/base/reducedim.jl#L1036) and ...

This is needed because Julia doesn't currently hoist the bounds checks in the most simple of next cases.  It's also limited to indexing into `::AbstractArray`.  It's what we demand of `AbstractArray` implementations.